### PR TITLE
Fix #5268: canonicalize directory, don't canonicalize file name

### LINKED ISSUE #5268: Termination error reporting crashes with symlinks
Scenario:
- A file `a/b/c/Foo.agda` with content
  ```agda
  S : Set
  S = S
  ```
- A symlink `c` pointing to `a/b/c`.
- Loading `c/Foo.agda` into emacs.
- `C-c C-l`.

Results in:
```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Errors.hs:295:18 
```
The crash occurs when termination errors as supposed to be printed at
https://github.com/agda/agda/blob/8d2c172cc2bc700e6eed993557ae11b19cedf682/src/full/Agda/TypeChecking/Errors.hs#L293-L295
and a module name for the current file is not found in `sourceToModule`.

I suppose I introduced this regression when fixing #4828.

### PR DESCRIPTION
**Pull request by uncle-betty**
This modifies the fix for #4828. Instead of not canonicalizing at all, canonicalize the directory but not the file name.

I think that this keeps the fix for #4828 intact while also fixing #5268.

### CODE REVIEW

**Review by andreasabel — APPROVED** [review]
Your solution sounds reasonable.

There might be other problems lurking with not-fully canonical names, though.  I hope we are not just pushing the frontier back...

**Comment by uncle-betty** [comment]
Thanks for the review!

Agreed - the fix is not the optimal solution, and it introduces not-fully canonical paths as yet another concept on top of the already existing paths, absolute paths, and canonical paths. This added complexity makes these things harder to reason about. (At least for me.)

Maybe we ultimately want a `CanonicalPath` type to properly separate the different worlds for everything - source code paths, include paths, ... - and get rid of not-fully canonical paths again.

I was just a little reluctant to make any major changes to the Agda code base *that* close to a major release. The risk management aspect of the proposed fix is what I found appealing about it. Other than that it's... a bit of a hack.

Maybe it's time for me to start punching above my weight after the 2.6.2 release and try my hand at more intrusive changes to the Agda code base like adding the `CanonicalPath` type. Maybe that would be a good next project for me.

File: src/full/Agda/Utils/FileName.hs [hunk 1]
```diff
@@ -74,11 +74,12 @@ absolute f = mkAbsolute <$> do
   -- canonicalizePath sometimes truncates paths pointing to
   -- non-existing files/directories.
   ex <- doesFileExist f `or2M` doesDirectoryExist f
-  if ex then
+  if ex then do
     -- Andreas, 2020-08-11, issue #4828
     -- Do not use @canonicalizePath@ here as it resolves symlinks,
     -- which leads to wrong placement of the .agdai file.
```
Comment (gallais) [inline src/full/Agda/Utils/FileName.hs]: If we do use `canonicalizePath` then we should update the comment.

Comment (uncle-betty) [reply src/full/Agda/Utils/FileName.hs]: True! I made the change and also rebased this branch onto current `master`.

**Pull request merged**